import "kernel/factors_selector/proto/factors_config.proto";

package NBuildModels.NProto;

// IMPORTANT: Using enums in model metainformation is strictly prohibited. For motivation see SEARCH-10049.


enum EApplyMode {
    ALWAYS = 0;
    WEB = 1;
    WEB_L2 = 9;
    WEB_EU_CHOICE_SCREEN = 7;
    ITD_ITP = 2;
    WEB_NN_OVER_FEATURES = 3;
    YA_MUSIC_FOR_ALICE = 5;
    NTM = 6;
    YABS_RT_USER = 10;
    YABS_RT = 11;
    ITDITP_NN_OVER_FEATURES = 12;
    ZEN_RT_USER = 13;
    ZEN_RT = 14;
    WEB_QUERY = 15;
    WEB_DOC = 16;
    WEB_QUERYDOC = 27;
    WEB_EARLYBINDING = 28;
    SEARCH_RT_USER = 17;
    SEARCH_RT_DOC = 18;
    WEB_NN_OVER_RAPID_CLICKS_FEATURES = 19;
    YABS_REQUEST_RT_USER = 20;
    MARKET_RT_DOC = 25;
    MARKET_RT_USER = 21;
    MARKET_RT = 22;
    TURBO_RT_USER = 23;
    TURBO_RT = 24;
    SHMICK = 26;
    WEB_ARAMUSIC_L2 = 29;
} //LAST_ID: 29

message TLayerInfo {
    optional string Name = 1;
    optional uint64 ExpectedSize = 2;

    // Next field is used only for external layers of NNs over features
    repeated NNeuralNet.NSelectFactors.NProto.TFactorsConfig InputFeatures = 3;
    optional bool Normalize = 4 [default = true];
}

message TParamInfo {
    optional string Name = 1;
    optional double Value = 2;
}

message TSubmodelInfo {
    optional string Name = 1;
    optional string Version = 2;

    // Next fields are filled while building optimized_model.dssm
    repeated TLayerInfo ExternalLayers = 3;
    repeated string OutputValues = 4;
    repeated TLayerInfo OutputLayers = 5;
    repeated TParamInfo OutputParameters = 6;
    optional EApplyMode ApplyMode = 7; // deprecated, see SEARCH-10049
    optional string ApplyConfiguration = 8;
}

message TFlagConfig {
    optional string Id = 1;
    optional bool Default = 2;
    repeated TSubmodelInfo UsedModels = 3;
}

message TModelsBundleConfig {
    repeated TSubmodelInfo SubmodelsInfos = 1;
    repeated TFlagConfig FlagsConfigs = 2;
    repeated string ModelsDirectories = 3;
}
